h2o-automl/src/main/java/ai/h2o/automl/TrainingStepsExecutor.java

```diff
@@ -0,0 +1,180 @@
+                }
+            }
+
+            try {
+                Thread.sleep(1000);
+            }
+            catch (InterruptedException e) {
+                // keep going
+            }
+            lastWorkedSoFar = workedSoFar;
+        }
+
+        // pick up any stragglers:
+        if (JobType.HyperparamSearch == work._type) {
+            if (job.isCrashed()) {
+                eventLog().warn(Stage.ModelTraining, jobDescription + " failed: " + job.ex().toString());
+            } else if (job.get() == null) {
+                eventLog().info(Stage.ModelTraining, jobDescription + " cancelled");
+            } else {
+                Grid<?> grid = (Grid) job.get();
+                int totalGridModelsBuilt = grid.getModelCount();
+                if (totalGridModelsBuilt > lastTotalGridModelsBuilt) {
+                    eventLog().debug(Stage.ModelTraining, "Built: " + totalGridModelsBuilt + " models for search: " + jobDescription);
+                    this.addModels(grid.getModelKeys());
+                }
+                eventLog().debug(Stage.ModelTraining, jobDescription + " complete");
+            }
+        } else if (JobType.ModelBuild == work._type) {
+            if (job.isCrashed()) {
+                eventLog().warn(Stage.ModelTraining, jobDescription + " failed: " + job.ex().toString());
+            } else if (job.get() == null) {
+                eventLog().info(Stage.ModelTraining, jobDescription + " cancelled");
+            } else {
+                eventLog().debug(Stage.ModelTraining, jobDescription + " complete");
+                this.addModel((Model) job.get());
+            }
+        }
+
+        // add remaining work
+        if (null != parentJob) {
+            parentJob.update(work._weight - lastWorkedSoFar);
+        }
+        work.consume();
+        _jobs.remove(job);
+    }
+
+
+    // If we have multiple AutoML engines running on the same project they will be updating the Leaderboard concurrently,
+    // so always use leaderboard() instead of the raw field, to get it from the DKV.
+    // Also, the leaderboard will reject duplicate models, so use the difference in Leaderboard length here.
+    private void addModels(final Key<Model>[] newModels) {
+        int before = leaderboard().getModelCount();
+        leaderboard().addModels(newModels);
+        int after = leaderboard().getModelCount();
+        _modelCount.addAndGet(after - before);
+    }
+
+    private void addModel(final Model newModel) {
+        int before = leaderboard().getModelCount();
+        leaderboard().addModel(newModel);
+        int after = leaderboard().getModelCount();
+        _modelCount.addAndGet(after - before);
+    }
+
+    private EventLog eventLog() {
+        return _eventLog == null ? null : (_eventLog = _eventLog._key.get());
+    }
+
+    private Leaderboard leaderboard() {
+        return _leaderboard == null ? null : (_leaderboard = _leaderboard._key.get());
```

Comment:
mmm, this null-check doesn't make sense anyway: `EventLog` and `Leaderboard` are injected through constructor and are `Iced` components (and therefore Serializable), so they can't be null.
However, we still need to ensure we always have a fresh instance, so I guess it can be simplified to:
```
private Leaderboard leaderboard() {
        return (_leaderboard = _leaderboard._key.get());
```

Reply:
in that case - my comment from above is most likely valid - leaderboard shoudln't be serialized a part of this class